ts/reviewer/image_occlusion.ts

```diff
@@ -0,0 +1,112 @@
+    const size = limitSize({ width: image.width, height: image.height });
+    canvas.width = size.width;
+    canvas.height = size.height;
+    const imageNaturalWidth = image.naturalWidth;
+    const canvasWidth = canvas.width;
+    let shapeScaler = 1;
+    if (imageNaturalWidth > canvasWidth) {
+        shapeScaler = canvasWidth / imageNaturalWidth;
+    } else {
+        shapeScaler = imageNaturalWidth / canvasWidth;
+    }
+    drawShapes(ctx, shapeScaler);
+}
+
+function drawShapes(ctx: CanvasRenderingContext2D, shapeScaler: number): void {
+    const activeCloze = document.querySelectorAll(".cloze");
+    const inActiveCloze = document.querySelectorAll(".cloze-inactive");
+
+    for (const clz of activeCloze) {
+        const cloze = (<HTMLDivElement> clz);
+        const shape = cloze.dataset.shape!;
+        const fill = cloze.dataset.questionmaskcolor!;
+        draw(ctx, cloze, shape, fill, shapeScaler);
+    }
+
+    for (const clz of inActiveCloze) {
+        const cloze = (<HTMLDivElement> clz);
+        const shape = cloze.dataset.shape!;
+        const fill = cloze.dataset.fill!;
+        const hideinactive = cloze.dataset.hideinactive == "true";
+        if (!hideinactive) {
+            draw(ctx, cloze, shape, fill, shapeScaler);
+        }
+    }
+}
+
+function draw(
+    ctx: CanvasRenderingContext2D,
+    cloze: HTMLDivElement,
+    shape: string,
+    color: string,
+    shapeScaler: number,
+): void {
+    ctx.fillStyle = color;
+
+    const post_left = parseFloat(cloze.dataset.left!) * shapeScaler;
+    const pos_top = parseFloat(cloze.dataset.top!) * shapeScaler;
+    const width = parseFloat(cloze.dataset.width!) * shapeScaler;
+    const height = parseFloat(cloze.dataset.height!) * shapeScaler;
+
+    switch (shape) {
+        case "rect":
+            {
+                ctx.fillRect(post_left, pos_top, width, height);
+            }
+            break;
+
+        case "ellipse":
+            {
+                const rx = parseFloat(cloze.dataset.rx!) * shapeScaler;
+                const ry = parseFloat(cloze.dataset.ry!) * shapeScaler;
+                ctx.beginPath();
+                ctx.ellipse(post_left, pos_top, rx, ry, 0, 0, Math.PI * 2, false);
+                ctx.fill();
+            }
+            break;
+
+        case "polygon":
+            {
+                const points = JSON.parse(cloze.dataset.points!);
+                ctx.beginPath();
+                ctx.moveTo(points[0][0] * shapeScaler, points[0][1] * shapeScaler);
+                for (let i = 1; i < points.length; i++) {
+                    ctx.lineTo(points[i][0] * shapeScaler, points[i][1] * shapeScaler);
+                }
+                ctx.closePath();
+                ctx.fill();
+            }
+            break;
+
+        default:
+            break;
+    }
+}
+
+function limitSize(size: { width: number; height: number }): { width: number; height: number; scalar: number } {
```

Comment:
Please add a comment noting this is copy+pasted from mask-editor.ts, and another in that file indicating that the code must be updated in this file as well, so we don't accidentally modify one but not the other.